### PR TITLE
Game refresher2

### PR DESCRIPTION
**Pull request by claudiociardelli**
Editor / Tools / Refresh Predefined Setups
Needs performance improvements, but limited by the low performane of PieceCloner.

I still need to see how to use the new ZipArchive methods.

Would like to have a review of the GameState.setup method. We had to add a check to exit the method when loading a Predefined Setup from the Editor to bypass the "Start new log file" and "Enter comments" popups.

### CODE REVIEW

**Comment by BrentEaston** [comment]
Hi Claudio, 

Need some Conflicts resolved.

I tried testing this on the GJS module, selected the new Update predefined setups option, clicked all options on and got the following crash:

2020-12-03 13:38:05,279 [0-main] INFO  VASSAL.launch.StartUp - Starting
2020-12-03 13:38:05,295 [0-main] INFO  VASSAL.launch.StartUp - OS Windows 10 10.0
2020-12-03 13:38:05,295 [0-main] INFO  VASSAL.launch.StartUp - Java version 14.0.1
2020-12-03 13:38:05,295 [0-main] INFO  VASSAL.launch.StartUp - VASSAL version 3.5.0-SNAPSHOT
2020-12-03 13:38:05,295 [0-main] INFO  VASSAL.launch.Launcher - Editor
2020-12-03 13:38:06,911 [0-AWT-EventQueue-0] WARN  VASSAL.i18n.BundleHelper - No Translation: Prefs.main_window
2020-12-03 13:38:09,016 [0-AWT-EventQueue-0] WARN  VASSAL.i18n.BundleHelper - No Translation: Deck.deck
2020-12-03 13:38:09,571 [0-AWT-EventQueue-0] INFO  VASSAL.build.GameModule - The Greatest Day: Sword, Juno and Gold Beaches version 1.8.9
2020-12-03 13:38:35,563 [0-AWT-EventQueue-0] INFO  VASSAL.configure.RefreshPredefinedSetupsDialog - Refresh Counters - Test Mode - Game will not be updated
2020-12-03 13:38:35,564 [0-AWT-EventQueue-0] INFO  VASSAL.configure.RefreshPredefinedSetupsDialog - 18 Predefined setups found
2020-12-03 13:38:35,565 [0-AWT-EventQueue-0] INFO  VASSAL.configure.RefreshPredefinedSetupsDialog - LE#1 Black Baron (BB-Black-Baron.vsav)
2020-12-03 13:38:35,566 [0-AWT-EventQueue-0] INFO  VASSAL.configure.RefreshPredefinedSetupsDialog - IN#1Storming Gold (IN1-Storming-Gold.vsav)
2020-12-03 13:38:35,566 [0-AWT-EventQueue-0] INFO  VASSAL.configure.RefreshPredefinedSetupsDialog - IN#2 Storming Juno (IN2-Storming-Juno.vsav)
2020-12-03 13:38:35,566 [0-AWT-EventQueue-0] INFO  VASSAL.configure.RefreshPredefinedSetupsDialog - IN#3 Storming Sword (IN3-Storming-Sword.vsav)
2020-12-03 13:38:35,567 [0-AWT-EventQueue-0] INFO  VASSAL.configure.RefreshPredefinedSetupsDialog - IN#4 To the Sea (IN4-To-The-Sea.vsav)
2020-12-03 13:38:35,568 [0-AWT-EventQueue-0] INFO  VASSAL.configure.RefreshPredefinedSetupsDialog - IN#5 Day of the Tiger (IN5-Day-of-the-Tiger.vsav)
2020-12-03 13:38:35,568 [0-AWT-EventQueue-0] INFO  VASSAL.configure.RefreshPredefinedSetupsDialog - IN#6 On to Bayeux (IN6-On-to-Bayeux.vsav)
2020-12-03 13:38:35,569 [0-AWT-EventQueue-0] INFO  VASSAL.configure.RefreshPredefinedSetupsDialog - IM#1 Day of Days (IM1-Day-of-Days.vsav)
2020-12-03 13:38:35,569 [0-AWT-EventQueue-0] INFO  VASSAL.configure.RefreshPredefinedSetupsDialog - IM#2 Race to Caen (IM2-Race-to-Caen.vsav)
2020-12-03 13:38:35,570 [0-AWT-EventQueue-0] INFO  VASSAL.configure.RefreshPredefinedSetupsDialog - IM#3 Saga of the 6th Airborne (IM3-Saga-of-the-6th-Airborne.vsav)
2020-12-03 13:38:35,570 [0-AWT-EventQueue-0] INFO  VASSAL.configure.RefreshPredefinedSetupsDialog - IM#4 O Canada (IM4-Oh-Canada.vsav)
2020-12-03 13:38:35,571 [0-AWT-EventQueue-0] INFO  VASSAL.configure.RefreshPredefinedSetupsDialog - IM#5 The Bloody battle of Tilly (IM5-The-Bloody-Battle-of-Tilly.vsav)
2020-12-03 13:38:35,571 [0-AWT-EventQueue-0] INFO  VASSAL.configure.RefreshPredefinedSetupsDialog - IM#6 Nemesis (IM6-Nemesis.vsav)
2020-12-03 13:38:35,572 [0-AWT-EventQueue-0] INFO  VASSAL.configure.RefreshPredefinedSetupsDialog - AD#1 The Right Hook (AD1-The-Right-Hook.vsav)
2020-12-03 13:38:35,572 [0-AWT-EventQueue-0] INFO  VASSAL.configure.RefreshPredefinedSetupsDialog - AD#2 The Initial Plan (AD2-The-Inital-Plan.vsav)
2020-12-03 13:38:35,572 [0-AWT-EventQueue-0] INFO  VASSAL.configure.RefreshPredefinedSetupsDialog - AD#3 Operation Perch (AD3-Operation-Perch.vsav)
2020-12-03 13:38:35,573 [0-AWT-EventQueue-0] INFO  VASSAL.configure.RefreshPredefinedSetupsDialog - CA#1 One Great Day (CA1-One-Great-Day.vsav)
2020-12-03 13:38:35,573 [0-AWT-EventQueue-0] INFO  VASSAL.configure.RefreshPredefinedSetupsDialog - CA#2 The Greatest Day: Sword, Juno and Gold Beach (CA2-The-Greatest-Day.vsav)
2020-12-03 13:38:35,574 [0-AWT-EventQueue-0] INFO  VASSAL.build.module.GameRefresher - ----------
2020-12-03 13:38:35,575 [0-AWT-EventQueue-0] INFO  VASSAL.build.module.GameRefresher - Updating Predefined Setup: LE#1 Black Baron ( BB-Black-Baron.vsav)
2020-12-03 13:38:35,860 [0-AWT-EventQueue-0] WARN  VASSAL.i18n.BundleHelper - No Translation: Deck.deck
2020-12-03 13:38:36,125 [0-AWT-EventQueue-0] INFO  VASSAL.script.ExpressionInterpreter - Attempting to load /VASSAL/script/init_expression.bsh URI generated=file:/J:/Data/git/vassal/vassal-app/target/classes/VASSAL/script/init_expression.bsh
2020-12-03 13:38:36,237 [0-AWT-EventQueue-0] INFO  VASSAL.build.module.GameRefresher - Collecting Counters
2020-12-03 13:38:36,238 [0-AWT-EventQueue-0] INFO  VASSAL.build.module.GameRefresher - 68 Counters found in game
2020-12-03 13:38:36,239 [0-AWT-EventQueue-0] INFO  VASSAL.build.module.GameRefresher - - 68 Counters collected
2020-12-03 13:38:36,239 [0-AWT-EventQueue-0] INFO  VASSAL.build.module.GameRefresher - - 0 Counters not collected - Not owned
2020-12-03 13:38:36,240 [0-AWT-EventQueue-0] INFO  VASSAL.build.module.GameRefresher - - 0 Counters not collected - Not visible
2020-12-03 13:38:36,240 [0-AWT-EventQueue-0] INFO  VASSAL.build.module.GameRefresher - -
2020-12-03 13:38:36,385 [0-AWT-EventQueue-0] ERROR VASSAL.tools.ErrorDialog - 
java.lang.ClassCastException: class VASSAL.counters.TriggerAction cannot be cast to class VASSAL.counters.Labeler (VASSAL.counters.TriggerAction and VASSAL.counters.Labeler are in unnamed module of loader 'app')
	at VASSAL.build.GpIdChecker.findState(GpIdChecker.java:341)
	at VASSAL.build.GpIdChecker.copyState(GpIdChecker.java:309)
	at VASSAL.build.GpIdChecker.createUpdatedPiece(GpIdChecker.java:245)
	at VASSAL.build.module.GameRefresher.processGamePiece(GameRefresher.java:272)
	at VASSAL.build.module.GameRefresher.execute(GameRefresher.java:251)
	at VASSAL.build.module.PredefinedSetup.refresh(PredefinedSetup.java:287)
	at VASSAL.configure.RefreshPredefinedSetupsDialog.refreshPredefinedSetups(RefreshPredefinedSetupsDialog.java:146)
	at VASSAL.configure.RefreshPredefinedSetupsDialog.lambda$initComponents$0(RefreshPredefinedSetupsDialog.java:65)
	at java.desktop/javax.swing.AbstractButton.fireActionPerformed(AbstractButton.java:1967)
	at java.desktop/javax.swing.AbstractButton$Handler.actionPerformed(AbstractButton.java:2308)
	at java.desktop/javax.swing.DefaultButtonModel.fireActionPerformed(DefaultButtonModel.java:405)
	at java.desktop/javax.swing.DefaultButtonModel.setPressed(DefaultButtonModel.java:262)
	at java.desktop/javax.swing.plaf.basic.BasicButtonListener.mouseReleased(BasicButtonListener.java:279)
	at java.desktop/java.awt.Component.processMouseEvent(Component.java:6636)
	at java.desktop/javax.swing.JComponent.processMouseEvent(JComponent.java:3342)
	at java.desktop/java.awt.Component.processEvent(Component.java:6401)
	at java.desktop/java.awt.Container.processEvent(Container.java:2263)
	at java.desktop/java.awt.Component.dispatchEventImpl(Component.java:5012)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2321)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4844)
	at java.desktop/java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4919)
	at java.desktop/java.awt.LightweightDispatcher.processMouseEvent(Container.java:4548)
	at java.desktop/java.awt.LightweightDispatcher.dispatchEvent(Container.java:4489)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2307)
	at java.desktop/java.awt.Window.dispatchEventImpl(Window.java:2764)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4844)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:772)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:715)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:391)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:95)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:745)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:743)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:391)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:742)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)

**Comment by claudiociardelli** [comment]
Brent, 
Shold work now. I was trying to cast OldPiece (GamePiece) as Labeler instead of its decorator.

File: vassal-app/src/main/java/VASSAL/build/GpIdChecker.java [hunk 1]
```diff
@@ -138,7 +158,7 @@ protected void testGpId(String id, SlotElement element) {
      *  If this has been called from a ModuleExtension, the GpId is prefixed with
      *  the Extension Id. Remove the Extension Id and just process the numeric part.
      *
-     *  NOTE: If GpIdChecker is being used by the GameRefresher, then there may be
+     *  NOTE: If GpIdChecker is being used by the GameRefesher, then there may be
```
Comment (BrentEaston) [inline vassal-app/src/main/java/VASSAL/build/GpIdChecker.java:161]: Typo on GameRefesher

**Review by BrentEaston — COMMENTED** [review]
On the whole. this is looking really good Claudio. 

There was a bit of flashing during the refresh of all predefined setups, but it was not terrible. It would be good to have some sort of progress dialog our in front telling you what is going on, because it isn't clear what it happening. However, I was not concerned about the speed of the refresh, I thought that was good. There is very little we can do to speed up the PieceCloner. As long as there is a good visual progress indicator, the speed is not an issue.

The Config dialogs need prettying up, but I will help with that when it is closer to done. 

I'm wondering whether the 2 new 'use names' options should default to ON? Is there a reason you would not want to use these? We will need extensive help to assist users to select the correct option to use.

**Comment by claudiociardelli** [comment]
Todos
Write doc
Player - add link to doc
ancient Update Saved Games tool- discuss about removal - not maintained and supposed to achieve same purpose. 
Editor - add progress indicator
Editor - chatter messages should be displayed as they are logged. currently appear at the end
Refresher shoudl update Global Properties states too.
Match Labelers/Layers by name. Remove checkboxes and make alwaus active?

**Comment by uckelman** [comment]
> ancient Update Saved Games tool- discuss about removal - not maintained and supposed to achieve same purpose.

This would be `SavedGameUpdater` and `SavedGameUpdaterDialog`?

**Pull request merged**